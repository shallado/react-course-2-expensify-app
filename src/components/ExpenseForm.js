import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends Component {
	state = {
		description: this.props.expense ? this.props.expense.description : '',
		note: this.props.expense ? this.props.expense.note : '',
		amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
		createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
		calendarFocused: false,
		error: ''
	};

	handleDescriptionChange = (e) => {
		const description = e.target.value;

		this.setState(() => ({ description }));
	};

	handleNoteChange = (e) => {
		const note = e.target.value;

		this.setState(() => ({ note }));
	};

	handleAmountChange = (e) => {
		const amount = e.target.value;

		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	};

	handleDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}
	};

	handleFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }));
	};

	handleSubmit = (e) => {
		e.preventDefault();
		let error;

		if (!this.state.description || !this.state.amount) {
			error = 'Please provide description and amount';
		} else {
			error = '';
			this.props.handleSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			});
		}
	
		this.setState(() => ({ error }));
	}

	render() {
		return (
			<div>
				{this.state.error && <p>{this.state.error}</p>}
				<form onSubmit={this.handleSubmit}>
					<input 
						type="text"
						placeholder="Description"
						autoFocus
						value={this.state.description}
						onChange={this.handleDescriptionChange}
					/>
					<input 
						type="text"
						placeholder="Amount"
						value={this.state.amount}
						onChange={this.handleAmountChange}
					/>
					<SingleDatePicker 
						date={this.state.createdAt}
						onDateChange={this.handleDateChange}
						focused={this.state.calendarFocused}
						onFocusChange={this.handleFocusChange}
						numberOfMonths={1}
						id="unique-id"
						isOutsideRange={() => false}
					/>
					<textarea
						placeholder="Add a note for your expense (optional)"
						value={this.state.note}
						onChange={this.handleNoteChange}
					>
					</textarea>
					<button>Add Expense</button>
				</form>
			</div>
		);
	}
}