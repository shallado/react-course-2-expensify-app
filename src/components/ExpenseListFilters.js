import React, { Component } from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends Component {
	state = {
		calendarFocused: null
	};

	handleDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	};

	handleFocusChange = (calendarFocused) => {
		this.setState(() => ({ calendarFocused }));
	};

	handleTextChange = (e) => {
		this.props.setTextFilter(e.target.value);
	}

	handleSortChange = (e) => {
		const sortValue = e.target.value;

		if (sortValue === 'date') {
			this.props.sortByDate();
		} else if (sortValue === 'amount') {
			this.props.sortByAmount();
		}
	}

	render() {
		return (
			<div>
				<input type="text" value={this.props.filters.text} onChange={this.handleTextChange} />
				<select value={this.props.filters.sortBy} onChange={this.handleSortChange}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.startDate}
					startDateId="start-date"
					endDate={this.props.filters.endDate}
					endDateId="end-date"
					onDatesChange={this.handleDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.handleFocusChange}
					showClearDates={true}
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

const mapStateToProps = (state) => ({
	filters: state.filters
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);