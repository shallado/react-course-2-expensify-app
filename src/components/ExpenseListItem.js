import React from 'react';
import { Link } from 'react-router-dom';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({ id, description, amount, createdAt, dispatch}) => (
	<div>
		<Link to={`/edit/${id}`}>
			<h3>{description}</h3>
		</Link>
		<p>Amount: {amount}</p>
		<p>CreatedAt: {createdAt}</p>
	</div>
);

export default ExpenseListItem;