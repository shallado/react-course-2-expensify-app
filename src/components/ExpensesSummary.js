import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenses }) => {
	const totalExpenses = expenses.length;
	const expenseText = totalExpenses === 1 ? 'expense' : 'expenses';
	const totalAmount = numeral(getExpensesTotal(expenses) / 100).format('$0,0.00');

	return (
		<div>
			<h1>{`Viewing ${totalExpenses} ${expenseText} totaling ${totalAmount}`}</h1>
		</div>		
	);
};

const mapStateToProps = (state) => ({
	expenses: state.expenses
});

export default connect(mapStateToProps)(ExpensesSummary);