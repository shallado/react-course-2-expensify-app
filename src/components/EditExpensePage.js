import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends Component {
	handleSubmit = (expense) => {
		this.props.editExpense(this.props.expense.id, expense);
		this.props.history.push('/');
	};

	handleRemoveExpense = () => {
		this.props.removeExpense({ id: this.props.expense.id });
		this.props.history.push('/');
	};

	render() {
		return (
			<div>
				<ExpenseForm
					expense={this.props.expense}
					handleSubmit={this.handleSubmit}
				/>
				<button onClick={this.handleRemoveExpense}>Remove</button>
			</div>
		);
	}
}

const mapDispatchToProps = (dispatch) => ({
	editExpense: (id, updates) => dispatch(editExpense(id, updates)),
	removeExpense: (id) => dispatch(removeExpense(id))
});

const mapStateToProps = (state, props) => ({
	expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);