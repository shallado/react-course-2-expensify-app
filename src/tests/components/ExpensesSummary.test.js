import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary correctly with one expense', () => {
	const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]}/>);

	expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary correctly with multiple expenses', () => {
	const wrapper = shallow(<ExpensesSummary expenses={expenses}/>);

	expect(wrapper).toMatchSnapshot();
});