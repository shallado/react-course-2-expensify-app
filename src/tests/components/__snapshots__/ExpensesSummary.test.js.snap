// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 3 expenses totaling $1,141.95
  </h1>
</div>
`;

exports[`should render ExpensesSummary correctly with one expense 1`] = `
<div>
  <h1>
    Viewing 1 expense totaling $1.95
  </h1>
</div>
`;
